File: games_service/app/watchdog.py
import asyncio
import contextlib
import logging
from datetime import datetime, timezone, timedelta

from sqlalchemy import select, or_

from .database import SessionLocal
from .models import Game, GameStatus, SideToMove
from .realtime import game_ws_manager
from .schemas import WsGameFinishedPayload
from .services import GameService, GameServiceError, build_game_detail, extract_move_data

LOGGER = logging.getLogger(__name__)
WATCHDOG_INTERVAL_SECONDS = 15
RECENT_MOVES_LIMIT = 120
ABANDONED_GAME_TIMEOUT_MINUTES = 10  # Delete games without second player after 10 minutes


class TimeoutWatchdog:
	def __init__(self) -> None:
		self._task: asyncio.Task | None = None

	def start(self) -> None:
		if self._task and not self._task.done():
			return
		self._task = asyncio.create_task(self._run(), name="timeout-watchdog")

	async def stop(self) -> None:
		if not self._task:
			return
		self._task.cancel()
		with contextlib.suppress(asyncio.CancelledError):
			await self._task
		self._task = None

	async def _run(self) -> None:
		while True:
			try:
				await self._tick()
			except asyncio.CancelledError:
				raise
			except Exception:  # pragma: no cover - defensive logging
				LOGGER.exception("Timeout watchdog iteration failed")
			await asyncio.sleep(WATCHDOG_INTERVAL_SECONDS)

	async def _tick(self) -> None:
		async with SessionLocal() as db:
			service = GameService(db)
			
			# 1. Check active games for timeout
			stmt = select(Game).where(Game.status == GameStatus.ACTIVE.value)
			result = await db.execute(stmt)
			active_games = result.scalars().all()
			for game in active_games:
				try:
					white_clock, black_clock = await service._compute_effective_clocks(game)
				except Exception:
					LOGGER.exception("Failed to compute clocks for game %s", game.id)
					continue

				loser: SideToMove | None = None
				if white_clock <= 0 and black_clock > 0:
					loser = SideToMove.WHITE
				elif black_clock <= 0 and white_clock > 0:
					loser = SideToMove.BLACK
				elif white_clock <= 0 and black_clock <= 0:
					loser = SideToMove.WHITE

				if not loser:
					continue

				requested_by = game.black_id if loser == SideToMove.WHITE else game.white_id
				if requested_by is None:
					continue

				try:
					finished_game = await service.timeout(
						game.id, loser_color=loser, requested_by=requested_by
					)
				except GameServiceError as exc:
					if exc.message not in {"White clock has not expired", "Black clock has not expired"}:
						LOGGER.warning(
							"Auto-timeout failed for game %s: %s", game.id, exc.message
						)
					continue

				moves = await service.get_moves(game.id, limit=RECENT_MOVES_LIMIT)
				# Извлекаем данные из Move объектов в async контексте
				move_data = [await extract_move_data(move) for move in moves]
				detail = build_game_detail(finished_game, moves=move_data)
				await game_ws_manager.broadcast(
					game.id,
					WsGameFinishedPayload(type="game_finished", game=detail).model_dump(mode="json"),
				)
			
			# 2. Clean up abandoned games (without second player for more than 10 minutes)
			cutoff_time = datetime.now(timezone.utc) - timedelta(minutes=ABANDONED_GAME_TIMEOUT_MINUTES)
			stmt = select(Game).where(
				Game.status == GameStatus.CREATED.value,
				Game.move_count == 0,
				or_(Game.white_id.is_(None), Game.black_id.is_(None)),
				Game.created_at < cutoff_time
			)
			result = await db.execute(stmt)
			abandoned_games = result.scalars().all()
			
			deleted_count = 0
			for game in abandoned_games:
				try:
					await db.delete(game)
					deleted_count += 1
					LOGGER.info("Deleted abandoned game %s (created at %s, missing player)", game.id, game.created_at)
					await game_ws_manager.broadcast(
						game.id,
						{
							"type": "game_cancelled",
							"game_id": str(game.id),
						},
					)
				except Exception:
					LOGGER.exception("Failed to delete abandoned game %s", game.id)
			
			if deleted_count > 0:
				await db.commit()
				LOGGER.info("Timeout watchdog deleted %d abandoned game(s)", deleted_count)


timeout_watchdog = TimeoutWatchdog()

File: games_service_go/internal/watchdog/watchdog.go
package watchdog

import (
	"context"
	"log"
	"time"

	"github.com/yourorg/games_service_go/internal/database"
	"github.com/yourorg/games_service_go/internal/models"
	"github.com/yourorg/games_service_go/internal/realtime"
	"github.com/yourorg/games_service_go/internal/services"
)

const (
	watchdogIntervalSeconds     = 15
	abandonedGameTimeoutMinutes = 60
	// Время ожидания первого хода для активных игр (в минутах)
	activeGameNoMoveTimeoutMinutes = 5
)

type Watchdog struct {
	db        *database.DB
	wsManager *realtime.ConnectionManager
	stopChan  chan struct{}
	service   *services.GameService
}

func New(db *database.DB, wsManager *realtime.ConnectionManager) *Watchdog {
	return &Watchdog{
		db:        db,
		wsManager: wsManager,
		stopChan:  make(chan struct{}),
		service:   services.NewGameService(db),
	}
}

func (w *Watchdog) Start() {
	ticker := time.NewTicker(watchdogIntervalSeconds * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			w.tick()
		case <-w.stopChan:
			return
		}
	}
}

func (w *Watchdog) Stop() {
	close(w.stopChan)
}

func (w *Watchdog) tick() {
	// Используем контекст с таймаутом для всех операций watchdog
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// 1. Проверка активных игр на таймаут
	w.checkTimeouts(ctx)

	// 2. Очистка заброшенных игр (CREATED без игроков)
	w.cleanupAbandonedGames(ctx)

	// 3. Очистка активных игр без ходов
	w.cleanupActiveGamesWithoutMoves(ctx)
}

func (w *Watchdog) checkTimeouts(ctx context.Context) {
	// turn_deadline_at is updated in the same transaction as a move, so the
	// watchdog can use an indexed query instead of scanning every active game.
	var games []models.Game
	now := time.Now().UTC()
	if err := w.db.WithContext(ctx).
		Where("status = ? AND turn_deadline_at IS NOT NULL AND turn_deadline_at <= ?", models.GameStatusActive, now).
		Order("turn_deadline_at ASC").
		Limit(100).
		Find(&games).Error; err != nil {
		log.Printf("[Watchdog] failed to query expired games: %v", err)
		return
	}

	for i := range games {
		loserColor := "white"
		if games[i].NextTurn == models.SideBlack {
			loserColor = "black"
		}
		if err := w.service.TimeoutAuto(ctx, games[i].ID, loserColor); err != nil {
			// Another request or replica can legitimately finish the game first.
			log.Printf("[Watchdog] failed to timeout game %s: %v", games[i].ID, err)
			continue
		}

		gameDetail, err := w.service.GetGame(ctx, games[i].ID)
		if err != nil {
			log.Printf("[Watchdog] failed to get game detail for %s: %v", games[i].ID, err)
			continue
		}
		_ = w.wsManager.Broadcast(games[i].ID, map[string]interface{}{
			"type":     "game_finished",
			"revision": gameDetail.MoveCount,
			"game":     gameDetail.Game,
		})
	}
}

func (w *Watchdog) cleanupAbandonedGames(ctx context.Context) {
	cutoffTime := time.Now().UTC().Add(-abandonedGameTimeoutMinutes * time.Minute)

	// Игра считается заброшенной, если:
	// 1. Статус = CREATED
	// 2. Создана более N минут назад (abandonedGameTimeoutMinutes)
	// 3. Не заполнены оба игрока (проверяем и user_id и session_id в metadata)
	//
	// Игра считается заполненной, если:
	// - (white_id IS NOT NULL OR metadata->>'white_session_id' IS NOT NULL) AND
	// - (black_id IS NOT NULL OR metadata->>'black_session_id' IS NOT NULL)
	//
	// Значит, игра заброшена, если:
	// - NOT (имеется white И имеется black)
	// = (white_id IS NULL AND COALESCE(metadata->>'white_session_id', '') = '') OR
	//   (black_id IS NULL AND COALESCE(metadata->>'black_session_id', '') = '')

	whereClause := `status = ? AND created_at < ? AND (
		(white_id IS NULL AND COALESCE(metadata->>'white_session_id', '') = '') OR
		(black_id IS NULL AND COALESCE(metadata->>'black_session_id', '') = '')
	)`

	var abandonedGames []models.Game
	if err := w.db.WithContext(ctx).Raw(
		"DELETE FROM games WHERE "+whereClause+" RETURNING id",
		models.GameStatusCreated, cutoffTime,
	).Scan(&abandonedGames).Error; err != nil {
		log.Printf("[Watchdog] failed to delete abandoned games: %v", err)
		return
	}

	// Broadcast отмены для каждой игры
	for _, game := range abandonedGames {
		w.wsManager.Broadcast(game.ID, map[string]interface{}{
			"type":    "game_cancelled",
			"game_id": game.ID.String(),
		})
	}
}

// cleanupActiveGamesWithoutMoves удаляет активные игры, где оба игрока присоединились,
// но никто не сделал первый ход в течение заданного времени
func (w *Watchdog) cleanupActiveGamesWithoutMoves(ctx context.Context) {
	cutoffTime := time.Now().UTC().Add(-activeGameNoMoveTimeoutMinutes * time.Minute)

	// Игра должна быть удалена, если:
	// 1. Статус = ACTIVE
	// 2. move_count = 0 (нет ходов)
	// 3. started_at < cutoffTime (прошло достаточно времени с момента старта)
	// 4. Оба игрока присоединились (проверяем через white_id/black_id или session_id)

	whereClause := `status = ? AND move_count = 0 AND started_at IS NOT NULL AND started_at < ? AND (
		(white_id IS NOT NULL OR COALESCE(metadata->>'white_session_id', '') != '') AND
		(black_id IS NOT NULL OR COALESCE(metadata->>'black_session_id', '') != '')
	)`

	var gamesToDelete []models.Game
	if err := w.db.WithContext(ctx).Raw(
		"DELETE FROM games WHERE "+whereClause+" RETURNING id",
		models.GameStatusActive, cutoffTime,
	).Scan(&gamesToDelete).Error; err != nil {
		log.Printf("[Watchdog] failed to delete active games without moves: %v", err)
		return
	}

	if len(gamesToDelete) > 0 {
		log.Printf("[Watchdog] deleted %d active games without moves", len(gamesToDelete))
	}

	// Broadcast отмены для каждой удаленной игры
	for _, game := range gamesToDelete {
		w.wsManager.Broadcast(game.ID, map[string]interface{}{
			"type":    "game_cancelled",
			"game_id": game.ID.String(),
			"reason":  "no_moves_timeout",
		})
	}
}
